### PR TITLE
Looks Very Solid Still

### PR DESCRIPTION
**Pull request by DiegoLHendrix**

### CODE REVIEW

File: datasheets/LVSS_Design_Document_Link.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+https://docs.google.com/document/d/1H9CgwCci4K7sW7PdY-aGkw6wP-7VwqDOHIir3eYked4/edit?usp=sharing
```
Comment (mjmagee991) [inline datasheets/LVSS_Design_Document_Link.txt]: Upload the actual PDF, not a link. Links can die, but the PDF won't

File: include/LVSS/LVSS.hpp
Comment (mjmagee991) [inline include/LVSS/LVSS.hpp:1]: Why did you add a folder inside include called LVSS?

File: include/LVSS/LVSS.hpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
+#ifndef _LVSS_
+#define _LVSS_
+
+#pragma once
```
Comment (mjmagee991) [inline include/LVSS/LVSS.hpp]: You need either the header guard or `#pragma once`, but you don't need both

File: include/LVSS/LVSS.hpp
Comment (mjmagee991) [reply include/LVSS/LVSS.hpp:1]: Also, your driver should be its own class that's used by the LVSS class. You shouldn't integrate your code directly into the LVSS class itself. This is like the rampup project, how we had the RampupBoard class that held an instance of the TMP class. Copy your code out of this file into a new current sensor class and then revert back to how it was before

File: src/LVSS.cpp
Comment (mjmagee991) [inline src/LVSS.cpp:1]: Like I said above, copy your code out of this into its own class and then revert this file

File: src/LVSS.cpp [hunk 3]
```diff
@@ -1,3 +1,34 @@
-#include <LVSS.hpp>
+#include <LVSS/LVSS.hpp>
 
-namespace LVSS {}
+namespace LVSS {
+
+LVSS::LVSS(IO::GPIO& lvssEn, IO::ADC& adc0) : LVSS_EN(lvssEn), ADC(adc0) {
+    //TODO: Implement
```
Comment (mjmagee991) [inline src/LVSS.cpp]: There's nothing more to be implemented here
```suggestion
```

File: src/LVSS.cpp [hunk 4]
```diff
@@ -1,3 +1,34 @@
...
+
+CO_OBJ_T* LVSS::getObjectDictionary() {
+    return nullptr;
+}
+
+void LVSS::process() {
+}
+
+uint32_t LVSS::readCurrent(IO::ADC& adc0) {
+    uint32_t current, voltage = static_cast<uint32_t>(adc0.read() * 1000);
```
Comment (mjmagee991) [inline src/LVSS.cpp]:
```suggestion
    uint32_t voltage = static_cast<uint32_t>(adc0.read() * 1000);
```

File: src/LVSS.cpp [hunk 5]
```diff
@@ -1,3 +1,34 @@
...
+    uint32_t r = (50 * 10 ^ -3), r2 = 5000, r3 = 50000;
```
Comment (mjmagee991) [inline src/LVSS.cpp]: Make all of these "magic numbers" named constants to match the comment below

File: src/LVSS.cpp [hunk 6]
```diff
@@ -1,3 +1,34 @@
...
+
+    //Current = (Vout * 5k ohms) / (Rshunt * R3)
+    current = ((voltage * r2) / (r * r3));
```
Comment (mjmagee991) [inline src/LVSS.cpp]: Move this down here for readability
```suggestion
    uint32_t current = ((voltage * r2) / (r * r3));
```

File: src/LVSS.cpp (hunk 5 above)
Comment (mjmagee991) [reply src/LVSS.cpp]: Also, don't instantiate multiple variables on one line

File: targets/REV3-LVSS/main.cpp
Comment (mjmagee991) [inline targets/REV3-LVSS/main.cpp:1]: Similar to what I was saying before, you don't want to modify this main file. Make a separate main file that just demonstrates how the current sensor driver works that doesn't have to deal with any other stuff like CAN

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Actual logic looks good, but needs some formatting work

File: include/LVSS/LVSS.hpp
Comment (EatonWu) [reply include/LVSS/LVSS.hpp:1]: > Why did you add a folder inside include called LVSS?

Replicating file structure from HUDL:

image

File: include/LVSS/LVSS.hpp (hunk 2 above)
Comment (EatonWu) [reply include/LVSS/LVSS.hpp]: ok dude

File: include/LVSS/LVSS.hpp
Comment (mjmagee991) [reply include/LVSS/LVSS.hpp:1]: That's fair. This was our directory structure a while back, but I changed it and I guess I missed the HUDL repo. I just created a PR to fix this on the HUDL. Sorry for the confusion

File: CMakeLists.txt [hunk 7]
```diff
@@ -57,7 +57,9 @@ project(${BOARD_LIB_NAME}
         LANGUAGES CXX C
         )
 
-add_library(${PROJECT_NAME} STATIC)
+add_library(${PROJECT_NAME} STATIC
```
Comment (mjmagee991) [inline CMakeLists.txt:60]: Revert this back to how it was. you shouldn't put these files here

File: CMakeLists.txt [hunk 8]
```diff
@@ -57,7 +57,9 @@ project(${BOARD_LIB_NAME}
...
+        include/INA138.hpp
+        src/INA138.cpp)
 
 # Add sources
 target_sources(${PROJECT_NAME} PRIVATE
```
Comment (mjmagee991) [inline CMakeLists.txt:64]: This is where you want to add the file, and you'll only want to put the src file
```suggestion
target_sources(${PROJECT_NAME} PRIVATE
        src/INA138.cpp
```

File: datasheets/Current_Sensing_High_Power.pdf
Comment (mjmagee991) [inline datasheets/Current_Sensing_High_Power.pdf:1]: Why are there 2 datasheets?

File: datasheets/REV3 LVSS - Official Documentation.txt
Comment (mjmagee991) [inline datasheets/REV3 LVSS - Official Documentation.txt:1]: This documentation is maintained by the team, so we don't need to store it in the repo. The reason I want to store the actual datasheets in the repo is so we still have them if the product gets removed from digikey

File: include/INA138.hpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
+#ifndef LVSS_INA138_HPP
+#define LVSS_INA138_HPP
+
+#include "EVT/dev/LCD.hpp"
```
Comment (mjmagee991) [inline include/INA138.hpp]: Use angle bracket includes for everything
```suggestion
#include <EVT/dev/LCD.hpp>
```
Also, you definitely don't need to include all these files. Remove the ones you don't need

File: include/INA138.hpp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    /**
+      * Calculates current using the input voltage rail into the Vicor
+      *
+      * @return the current
+      */
+    uint32_t readCurrent(IO::ADC&);
+
+private:
+    //ADC variable for getting input voltage
```
Comment (mjmagee991) [inline include/INA138.hpp]:
```suggestion
    /** ADC instance for getting input voltage */
```

File: include/INA138.hpp [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+public:
+    /**
+      * Constructor for current sensing class
+      */
+    INA138(IO::ADC& adc0);
+
+    /**
+      * Calculates current using the input voltage rail into the Vicor
+      *
+      * @return the current
```
Comment (mjmagee991) [inline include/INA138.hpp]:
```suggestion
      * @return The current
```

File: include/INA138.hpp [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+ */
+class INA138 {
+public:
+    /**
+      * Constructor for current sensing class
+      */
+    INA138(IO::ADC& adc0);
+
+    /**
+      * Calculates current using the input voltage rail into the Vicor
```
Comment (mjmagee991) [inline include/INA138.hpp]: Don't make your comments specific to the design of the hardware. This driver is just for an IC that reads current. You can include more detail in the LVSS class.
```suggestion
      * Get the current detected by the INA
```

File: include/INA138.hpp [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+#include "LVSS.hpp"
+
+namespace IO = EVT::core::IO;
+namespace DEV = EVT::core::DEV;
+namespace log = EVT::core::log;
+
+namespace LVSS {
+
+/**
+ * class for LVSS current sensor feature
```
Comment (mjmagee991) [inline include/INA138.hpp]:
```suggestion
 * Class for LVSS current sensor feature
```

File: targets/REV3-LVSS/main.cpp
Comment (mjmagee991) [inline targets/REV3-LVSS/main.cpp:1]: Revert this file to the way it was before. This PR should just be for stuff related to the INA. You should make a separate target to demonstrate the INA class

File: src/INA138.cpp [hunk 14]
```diff
@@ -0,0 +1,22 @@
+#include "INA138.hpp"
```
Comment (mjmagee991) [inline src/INA138.cpp]:
```suggestion
#include <INA138.hpp>
```

File: include/INA138.hpp
Comment (mjmagee991) [inline include/INA138.hpp:1]: Move this into a folder `include/dev`

File: src/INA138.cpp
Comment (mjmagee991) [inline src/INA138.cpp:1]: Move this into a folder `src/dev`

File: include/LVSS.hpp
Comment (mjmagee991) [inline include/LVSS.hpp:1]: I'm now realizing you have some of Eaton's changes in here because his branch was merged into yours. If I've mentioned anything that was done by Eaton and isn't related to your stuff, you can just ignore it, and I'll re-review it in his PR

File: src/INA138.cpp [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+namespace LVSS {
+
+INA138::INA138(IO::ADC& adc0) : ADC(adc0) {
```
Comment (mjmagee991) [inline src/INA138.cpp]:
```suggestion
INA138::INA138(IO::ADC& adc0) : ADC(adc0) {}
```

File: src/INA138.cpp [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+}
```
Comment (mjmagee991) [inline src/INA138.cpp]:
```suggestion
```

File: src/INA138.cpp [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+
+uint32_t INA138::readCurrent(IO::ADC& adc0) {
+    //Gets voltage from adc
+    uint32_t voltage = static_cast<uint32_t>(adc0.read() * 1000);
```
Comment (mjmagee991) [inline src/INA138.cpp]: Gets voltage as a number from 0-4095
```suggestion
    uint32_t adcCounts = adc0.readRaw();
```

File: include/INA138.hpp [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+      */
+    uint32_t readCurrent(IO::ADC&);
+
+private:
+    //ADC variable for getting input voltage
+    IO::ADC& ADC;
+
+    ///Read INA138 datasheet page 3
+    //Resistor variables
+    static constexpr uint32_t fixedPoint = 1000;         //Use fixpoint variable to multiply everything by 1k
```
Comment (mjmagee991) [inline include/INA138.hpp]: You don't actually want to just multiply everything by 1000. I'm going to modify the below to make it work, and then I'll explain it to you next meeting
```suggestion
```

File: include/INA138.hpp [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+    static constexpr uint32_t rShunt = 0.05 * fixedPoint;//0.5 ohm resistor
```
Comment (mjmagee991) [inline include/INA138.hpp]: Multiply this by just enough to make to make it an integer
```suggestion
    static constexpr uint32_t rShunt = 0.05 * 100; //0.5 ohm resistor
```

File: include/INA138.hpp [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    static constexpr uint32_t r1 = 5000 * fixedPoint;    //5k ohm
```
Comment (mjmagee991) [inline include/INA138.hpp]:
```suggestion
    static constexpr uint32_t r1 = 5000t;    //5k ohm
```

File: include/INA138.hpp [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+    static constexpr uint32_t r3 = 50000 * fixedPoint;   //50k ohm
```
Comment (mjmagee991) [inline include/INA138.hpp]:
```suggestion
    static constexpr uint32_t r3 = 50000;   //50k ohm
```

File: src/INA138.cpp [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+    voltage *= fixedPoint;
```
Comment (mjmagee991) [inline src/INA138.cpp]: Don't need to multiply the input by anything because the raw value is already an integer
```suggestion
```

File: src/INA138.cpp [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+
+    //Current = (Vout * 5k ohms) / (Rshunt * R3)
+    uint32_t current = ((voltage * r1) / (rShunt * r3));
```
Comment (mjmagee991) [inline src/INA138.cpp]:
```suggestion
    uint32_t current = ((adcCounts * r1 * 330) / (rShunt * r3 * 4096));
```

File: src/INA138.cpp [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+INA138::INA138(IO::ADC& adc0) : ADC(adc0) {
+}
+
+uint32_t INA138::readCurrent(IO::ADC& adc0) {
+    //Gets voltage from adc
+    uint32_t voltage = static_cast<uint32_t>(adc0.read() * 1000);
+    voltage *= fixedPoint;
+
+    //Current = (Vout * 5k ohms) / (Rshunt * R3)
```
Comment (mjmagee991) [inline src/INA138.cpp]:
```suggestion
    // Current = ((adcCounts * 3.3) * 5k ohms * 100) / ((Rshunt * 100) * R3 * 4096)
```

File: src/INA138.cpp [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+    uint32_t current = ((voltage * r1) / (rShunt * r3));
+
+    //divide the current by 1k fixed point get the actual number
+    current /= fixedPoint;
```
Comment (mjmagee991) [inline src/INA138.cpp]:
```suggestion
```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
Getting close but needs more cleanup

File: CMakeLists.txt [hunk 26]
```diff
@@ -57,11 +57,14 @@ project(${BOARD_LIB_NAME}
...
 
-add_library(${PROJECT_NAME} STATIC)
+add_library(${PROJECT_NAME} STATIC
+        )
 
 # Add sources
 target_sources(${PROJECT_NAME} PRIVATE
         src/LVSS.cpp
+        src/dev/INA138.cpp
+        targets/REV3-INA138/INA138_main.cpp
```
Comment (mjmagee991) [inline CMakeLists.txt]: This should only have source files, not targets
```suggestion
```

File: include/LVSS.hpp [hunk 27]
```diff
@@ -1,13 +1,59 @@
-#pragma once
+#ifndef _LVSS_
+#define _LVSS_
+
+#include "EVT/dev/LCD.hpp"
```
Comment (mjmagee991) [inline include/LVSS.hpp]: Use angle brackets for all your includes
```suggestion
#include <EVT/dev/LCD.hpp>
```

File: include/LVSS.hpp [hunk 28]
```diff
@@ -1,13 +1,59 @@
...
+    static constexpr uint8_t NODE_ID = 42;
+
+    /**
+     * Placeholder constructor for the LVSS class
+     */
+    LVSS(IO::GPIO& lvssEn);
+
+    CO_OBJ_T* getObjectDictionary() override;
+
+    /**
```
Comment (mjmagee991) [inline include/LVSS.hpp]: Because these methods are overridden, they'll inherit Doxygen from their parent class. Remove the comments for this and the following function

File: include/dev/INA138.hpp [hunk 29]
```diff
@@ -0,0 +1,43 @@
+#ifndef LVSS_INA138_HPP
+#define LVSS_INA138_HPP
+
+#include "EVT/io/ADC.hpp"
```
Comment (mjmagee991) [inline include/dev/INA138.hpp]:
```suggestion
#include <EVT/io/ADC.hpp>
```

File: include/dev/INA138.hpp [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+#include "EVT/utils/log.hpp"
+
+namespace IO = EVT::core::IO;
+namespace DEV = EVT::core::DEV;
+namespace log = EVT::core::log;
+
+namespace LVSS {
+
+/**
+ * Class for LVSS current sensor feature
```
Comment (mjmagee991) [inline include/dev/INA138.hpp]:
```suggestion
 * Class for LVSS current sensor
```

File: include/dev/INA138.hpp [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+ */
+class INA138 {
+public:
+    /**
+      * Constructor for current sensing class
+      */
+    INA138(IO::ADC& adc0);
+
+    /**
+      * * Get the current detected by the INA
```
Comment (mjmagee991) [inline include/dev/INA138.hpp]:
```suggestion
      * Get the current detected by the INA
```

File: include/dev/INA138.hpp [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+      *
+      * @return The current
+      */
+    uint32_t readCurrent(IO::ADC&);
```
Comment (mjmagee991) [inline include/dev/INA138.hpp]: You should use the instance variable here instead of taking the ADC as an argument
```suggestion
    uint32_t readCurrent();
```

File: src/dev/INA138.cpp [hunk 33]
```diff
@@ -0,0 +1,17 @@
+#include "dev/INA138.hpp"
```
Comment (mjmagee991) [inline src/dev/INA138.cpp]:
```suggestion
#include <dev/INA138.hpp>
```

File: src/dev/INA138.cpp [hunk 34]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace LVSS {
+
+INA138::INA138(IO::ADC& adc0) : ADC(adc0) {}
+
+uint32_t INA138::readCurrent(IO::ADC& adc0) {
```
Comment (mjmagee991) [inline src/dev/INA138.cpp]:
```suggestion
uint32_t INA138::readCurrent() {
```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
A couple minor things to clean up, but this looks great!

**Review by EatonWu — APPROVED** [review]
Looks good, probably

File: include/dev/INA138.hpp [hunk 35]
```diff
@@ -0,0 +1,43 @@
...
+
+private:
+    /** ADC instance for getting input voltage */
+    IO::ADC& ADC;
+
+    ///Read INA138 datasheet page 3
+    //Resistor variables
+    static constexpr uint32_t rShunt = 0.05 * 100;//0.05 ohm resistor
+    static constexpr uint32_t r1 = 5000;          //5k ohm
+    static constexpr uint32_t r3 = 50000;         //50k ohm
```
Comment (AarC10) [inline include/dev/INA138.hpp]: Fix comment formatting

File: include/dev/INA138.hpp [hunk 36]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+}// namespace LVSS
+
+#endif//LVSS_INA138_HPP
```
Comment (AarC10) [inline include/dev/INA138.hpp]: Newline at EOF

File: targets/REV3-LVSS/main.cpp [hunk 37]
```diff
@@ -4,18 +4,98 @@
...
+ * function each time a new CAN message comes in.
+ *
+ * NOTE: For this sample, every non-extended (so 11  bit CAN IDs) will be
+ * assumed to be intended to be passed as a CANopen message.
+ *
+ * @param message[in] The passed in CAN message that was read.
+ */
+// create a can interrupt handler
+void canInterrupt(IO::CANMessage& message, void* priv) {
+    auto* queue = (types::FixedQueue<CANOPEN_QUEUE_SIZE, IO::CANMessage>*) priv;
```
Comment (AarC10) [inline targets/REV3-LVSS/main.cpp]: static_cast or reinterpret_cast this

**Review by AarC10 — DISMISSED** [review]

**Review by EatonWu — APPROVED** [review]

**Comment by mjmagee991** [comment]
[REV3 LVSS Current Sensor Driver](https://trello.com/c/luZr4z98/290-rev3-lvss-current-sensor-driver)

File: include/dev/INA138.hpp [hunk 38]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace LVSS {
+
+/**
+ * Class for LVSS current sensor
+ */
+class INA138 {
+public:
+    /**
+      * Constructor for current sensing class
```
Comment (mjmagee991) [inline include/dev/INA138.hpp]: This is the smallest of nitpicks, but your comment blocks are off by one space
```suggestion
     * Constructor for current sensing class
```

File: targets/REV3-INA138/INA138_main.cpp [hunk 39]
```diff
@@ -0,0 +1,38 @@
+/**
+* This is a basic sample to show how the current sensing feature of the
```
Comment (mjmagee991) [inline targets/REV3-INA138/INA138_main.cpp]:
```suggestion
 * This is a basic sample to show how the current sensing feature of the
```

File: targets/REV3-INA138/INA138_main.cpp [hunk 40]
```diff
@@ -0,0 +1,38 @@
...
+ * LVSS using the INA138 IC
+*/
```
Comment (mjmagee991) [inline targets/REV3-INA138/INA138_main.cpp]:
```suggestion
 */
```

File: targets/REV3-INA138/INA138_main.cpp [hunk 41]
```diff
@@ -0,0 +1,38 @@
...
+#include <EVT/utils/log.hpp>
+#include <EVT/utils/time.hpp>
+#include <LVSS.hpp>
+
+namespace IO = EVT::core::IO;
+namespace DEV = EVT::core::DEV;
+namespace time = EVT::core::time;
+namespace log = EVT::core::log;
+namespace types = EVT::core::types;
+using namespace std;
```
Comment (mjmagee991) [inline targets/REV3-INA138/INA138_main.cpp]: `using namespace` can cause a lot of issues, so you generally don't want to do that
```suggestion
```

File: targets/REV3-INA138/INA138_main.cpp [hunk 42]
```diff
@@ -0,0 +1,38 @@
...
+#include <EVT/manager.hpp>
+#include <EVT/utils/log.hpp>
+#include <EVT/utils/time.hpp>
+#include <LVSS.hpp>
+
+namespace IO = EVT::core::IO;
+namespace DEV = EVT::core::DEV;
+namespace time = EVT::core::time;
+namespace log = EVT::core::log;
+namespace types = EVT::core::types;
```
Comment (mjmagee991) [inline targets/REV3-INA138/INA138_main.cpp]: This isn't used
```suggestion
```

**Review by mjmagee991 — APPROVED** [review]
A couple nitpicks, but otherwise looks ready to go. Nice work!
